### PR TITLE
agent: optimize the code of systemd cgroup manager

### PR DESCRIPTION
**Pull request by Yuan-Zhuo**
1. Directly support CgroupManager::freeze through systemd API.
2. Avoid always passing unit_name by storing it into DBusClient.
3. Realize CgroupManager::destroy more accurately by killing systemd unit rather than stop it.
4. Ignore no such unit error when destroying systemd unit.
5. Update zbus version and corresponding interface file.

Acknowledgement: error handling for no such systemd unit error refers to https://github.com/kata-containers/kata-containers/pull/7081.

Fixes: #7080, #7142, #7143, #7166

### CODE REVIEW

File: src/agent/rustjail/src/cgroups/systemd/manager.rs [hunk 1]
```diff
@@ -29,7 +29,7 @@ pub struct Manager {
     pub mounts: HashMap<String, String>,
     pub cgroups_path: CgroupsPath,
     pub cpath: String,
-    pub unit_name: String,
+    // pub unit_name: String,
```
[the remark this answers is not in the document]
Comment (Yuan-Zhuo) [reply src/agent/rustjail/src/cgroups/systemd/manager.rs]: Fixed

File: src/agent/rustjail/src/cgroups/systemd/dbus_client.rs [hunk 2]
```diff
@@ -68,35 +62,65 @@ impl SystemdInterface for DBusClient {
...
+            .kill_unit(&self.unit_name, WHO_ENUM_ALL, SIGNAL_KILL)
+            .or_else(|e| match e {
+                zbus::Error::MethodError(error_name, _, _)
+                    if error_name.as_str() == NO_SUCH_UNIT_ERROR =>
+                {
+                    Ok(())
+                }
+                _ => Err(e),
+            })
+            .with_context(|| "failed to get systemd version".to_string())?;
```
[the remark this answers is not in the document]
Comment (Yuan-Zhuo) [reply src/agent/rustjail/src/cgroups/systemd/dbus_client.rs]: Fixed

File: src/agent/rustjail/src/cgroups/systemd/interface/system.rs [hunk 3]
```diff
@@ -1,4 +1,4 @@
-// Copyright 2021-2022 Kata Contributors
+// Copyright 2022-2023 Kata Contributors
```
[the remark this answers is not in the document]
Comment (Yuan-Zhuo) [reply src/agent/rustjail/src/cgroups/systemd/interface/system.rs]: Fixed

**Review by Tim-Zhang — APPROVED** [review]
LGTM, thanks @Yuan-Zhuo

**Comment by Amulyam24** [comment]
Hi @Yuan-Zhuo. Thank you for the PR. The zbus version update also fixes the error mentioned in this [issue](https://github.com/kata-containers/kata-containers/issues/7166). Do you mind appending the issue number to the `Fixes:` in your commit?

**Comment by Yuan-Zhuo** [comment]
> Hi @Yuan-Zhuo. Thank you for the PR. The zbus version update also fixes the error mentioned in this [issue](https://github.com/kata-containers/kata-containers/issues/7166). Do you mind appending the issue number to the `Fixes:` in your commit?

That's great, what a surprise!

**Comment by snir911** [comment]
@Yuan-Zhuo @yoheiueda I noticed that sometimes remove_container requests of an init-container are lengthening up to 40 seconds, do you have any idea if it may be related to the work being done here? i wonder if it may be the stop_unit that is blocking the call in certain cases for too long due to the missing unit, or something else/similar maybe.. WDYT?
* I'm also getting the "no such unit" error frequently
* my setup is CoCo on Azure with RHEL cvm
* adding this PR didn't help

**Comment by Amulyam24** [comment]
Hi @Yuan-Zhuo @Tim-Zhang, could you please share any time estimate for this PR to be merged? we depend on fixes from this PR for a working agent.

**Comment by Yuan-Zhuo** [comment]
> @Yuan-Zhuo @yoheiueda I noticed that sometimes remove_container requests of an init-container are lengthening up to 40 seconds, do you have any idea if it may be related to the work being done here? i wonder if it may be the stop_unit that is blocking the call in certain cases for too long due to the missing unit, or something else/similar maybe.. WDYT?
> 
> * I'm also getting the "no such unit" error pretty frequently
> * adding this PR didn't help
> * my setup is CoCo on Azure with RHEL cvm

This is an interesting problem, I will try solving it after this pr is merged.

**Comment by Yuan-Zhuo** [comment]
> @Yuan-Zhuo @yoheiueda I noticed that sometimes remove_container requests of an init-container are lengthening up to 40 seconds, do you have any idea if it may be related to the work being done here? i wonder if it may be the stop_unit that is blocking the call in certain cases for too long due to the missing unit, or something else/similar maybe.. WDYT?
> 
> * I'm also getting the "no such unit" error pretty frequently
> * adding this PR didn't help
> * my setup is CoCo on Azure with RHEL cvm

This is an interesting problem, I will try to solve it.

**Comment by stevenhorsman** [comment]
@Yuan-Zhuo - I wondered how this PR is getting on as I see if fixes quite a lot of issues, so would be great to get merged soon! It looks like it needs a rebase, but is there anything outstanding? If so would it be worth splitting up so some of the great fixes can getting in before then? Thanks!

**Comment by Yuan-Zhuo** [comment]
@Amulyam24 @stevenhorsman Thanks for your attention to this PR! I will resolve the conflict and push the PR to be merged as soon as possible.

**Review by lifupan — APPROVED** [review]

**Comment by stevenhorsman** [comment]
The WIP test was stuck, so I added and removed the label to re-trigger it

**Pull request merged**